src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -754,106 +759,247 @@ public static boolean bothSingleByteOptimizable(DynamicObject string, DynamicObj
     }
 
     @CoreMethod(names = "count", rest = true)
-    @ImportStatic(StringGuards.class)
     public abstract static class CountNode extends CoreMethodArrayArgumentsNode {
 
-        @Child private EncodingNodes.CheckEncodingNode checkEncodingNode = EncodingNodesFactory.CheckEncodingNodeGen.create(null, null);
         @Child private ToStrNode toStr = ToStrNodeGen.create(null);
+        @Child private CountRopesNode countRopesNode = CountRopesNode.create();
+
+        @Specialization(guards = "args.length == size", limit = "getCacheLimit()")
+        public int count(VirtualFrame frame, DynamicObject string, Object[] args,
+                @Cached("args.length") int size) {
+            final Rope[] ropes = argRopes(frame, args);
+            return countRopesNode.executeCount(string, ropes);
+        }
+
+        @Specialization(replaces = "count")
+        public int countSlow(VirtualFrame frame, DynamicObject string, Object[] args) {
+            final Rope[] ropes = argRopesSlow(frame, args);
+            return countRopesNode.executeCount(string, ropes);
+        }
+
+        @ExplodeLoop
+        protected Rope[] argRopes(VirtualFrame frame, Object[] args) {
+            final Rope[] strs = new Rope[args.length];
+            for (int i = 0; i < args.length; i++) {
+                strs[i] = rope(toStr.executeToStr(frame, args[i]));
+            }
+            return strs;
+        }
+
+        protected Rope[] argRopesSlow(VirtualFrame frame, Object[] args) {
+            final Rope[] strs = new Rope[args.length];
+            for (int i = 0; i < args.length; i++) {
+                strs[i] = rope(toStr.executeToStr(frame, args[i]));
+            }
+            return strs;
+        }
```

Comment:
Starts to look like we should write this in Ruby (we would need to profile the size in Array#map and explode if it's small) and add a primitive to get the rope out of a String.

Reply:
I looked at profiling for size when I was fixing up the array builder, but didn't get something I was happy with for both warm up and peak performance. I think it would be worth looking at again though because getting ropes out before calling something more primitive is a common pattern in the Java part of our run time.